syft/pkg/cataloger/java/graalvm_native_image_cataloger.go

```diff
@@ -0,0 +1,316 @@
+		return nil, err
+	}
+	dataSectionBase := dataSection.SectionHeader.Addr
+	data, err := dataSection.Data()
+	if err != nil {
+		log.Warnf("native-image cataloger: cannot read the .data section.")
+		return nil, err
+	}
+	sbomLocation := sbom.Value - dataSectionBase
+	lengthLocation := sbomLength.Value - dataSectionBase
+
+	return decompressSbom(data, sbomLocation, lengthLocation)
+}
+
+// FetchPkgs obtains the packages from a Native Image given as a Mach O file.
+func (ni NativeImageMachO) FetchPkgs() ([]pkg.Package, error) {
+	var sbom macho.Symbol
+	var sbomLength macho.Symbol
+	var svmVersion macho.Symbol
+
+	bi := ni.File
+	for _, s := range bi.Symtab.Syms {
+		switch s.Name {
+		case "_" + nativeImageSbomSymbol:
+			sbom = s
+		case "_" + nativeImageSbomLengthSymbol:
+			sbomLength = s
+		case "_" + nativeImageSbomVersionSymbol:
+			svmVersion = s
+		}
+	}
+	if sbom.Value == 0 || sbomLength.Value == 0 || svmVersion.Value == 0 {
+		log.Infof("native-image cataloger: %v.", nativeImageMissingSymbolsError)
+		return nil, errors.New(nativeImageMissingSymbolsError)
+	}
+
+	dataSegment := bi.Segment("__DATA")
+	if dataSegment == nil {
+		return nil, nil
+	}
+	databuf, err := dataSegment.Data()
+	if err != nil {
+		log.Warnf("native-image cataloger: cannot obtain buffer from data segment.")
+		return nil, nil
+	}
+	sbomLocation := sbom.Value - dataSegment.Addr
+	lengthLocation := sbomLength.Value - dataSegment.Addr
+
+	return decompressSbom(databuf, sbomLocation, lengthLocation)
+}
+
+// fetchPkgs provides the packages available in a UnionReader.
+func fetchPkgs(reader unionreader.UnionReader, filename string) []pkg.Package {
+	var pkgs []pkg.Package
+	imageformats := []func(string, io.ReaderAt) (NativeImage, error){newElf, newMachO}
+
+	// NOTE: multiple readers are returned to cover universal binaries, which are files
+	// with more than one binary
+	readers, err := unionreader.GetReaders(reader)
+	if err != nil {
+		log.Warnf("native-image cataloger: failed to open a binary: %v", err)
+		return nil
+	}
+	for _, r := range readers {
+		for _, make := range imageformats {
+			ni, err := make(filename, r)
+			if err != nil {
+				continue
+			}
+			newpkgs, err := ni.FetchPkgs()
+			if err != nil {
+				log.Warnf("native-image cataloger: error extracting SBOM from %s %v", filename, err)
+				continue
+			}
+			pkgs = append(pkgs, newpkgs...)
+		}
+	}
+	return pkgs
+}
+
+// Catalog attempts to find any native image executables reachable from a resolver.
+func (c *NativeImageCataloger) Catalog(resolver source.FileResolver) ([]pkg.Package, []artifact.Relationship, error) {
+	var pkgs []pkg.Package
+	fileMatches, err := resolver.FilesByMIMEType(internal.ExecutableMIMETypeSet.List()...)
+	if err != nil {
+		return pkgs, nil, fmt.Errorf("failed to find binaries by mime types: %w", err)
+	}
+
+	for _, location := range fileMatches {
+		readerCloser, err := resolver.FileContentsByLocation(location)
```

Comment:
The only other concern I have with this change is that it looks like every executable file is going to be read, contents loaded and scanned. I wonder how much this could slow down scans for large images.

Is there other discriminator we could use to identify "this is a graalvm  file" before reading all the contents?

Reply:
I can see the concern here -- the short answer is that I'm not certain if this will be impactful "enough" to take action or not. I don't know of an easier way to do the upfront "is this a graalvm file" check than what is done here since a graalvm file produces native binaries.

Reply:
I did an initial test with a few images and so far the difference run-time-wise is pretty nominal. Right now I don't have a large concern with this (could change in the future though).

Reply:
It looks like Go's packages for consuming binary files (i.e., ELF, Mach-O, and PE) use a `ReaderAt` to read data from the file. I think this implies that the Native Image cataloger will only read parts of the file relevant for detecting a valid executable (i.e. a correct header) and GraalVM Native Image symbols. That is, if the GraalVM version symbol is absent, the cataloger will stop reading the file.

Below, I `strace` `syft` while fetching the GraalVM SBOM from `ls` and a Native Image application. The cataloger appears to read the first part of `ls`, and then enumerates the symbol table. The cataloger reads less than the full size of the `ls` binary (which is 140kb) before concluding that there is no SBOM. As expected, the cataloger reads more data when reading the SBOM from the Native Image.

Based on this small experiment, it looks like you could expect our cataloger to make <50 small reads on each valid executable binary file in the container image. Of course, we can try to implement a more optimal check if this overhead is too high.

ls:

    openat(AT_FDCWD, "/path/to/test-container/ls", O_RDONLY|O_CLOEXEC) = 6
    pread64(6, "\177ELF", 4, 0)             = 4
    pread64(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0", 16, 0) = 16
    pread64(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0", 16, 0) = 16
    pread64(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\320g\0\0\0\0\0\0"..., 64, 0) = 64 x 43
    pread64(6, "\0.shstrtab\0.interp\0.note.gnu.pro"..., 285, 139932) = 285
    pread64(6, "\260h\0\0\0\0\0\0ph\0\0\0\0\0\0@s\0\0\0\0\0\0\240x\0\0\0\0\0\0"..., 4696, 135184) = 4696
    close(6)                                = 0
   
Native Image:

    openat(AT_FDCWD, "/path/to/test-container/native-application", O_RDONLY|O_CLOEXEC) = 6
    pread64(6, "\177ELF", 4, 0)             = 4
    pread64(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0", 16, 0) = 16
    pread64(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0", 16, 0) = 16
    pread64(6, "\177ELF\2\1\1\0\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\200\2W\0\0\0\0\0"..., 64, 0) = 64 x 39
    pread64(6, "\0.symtab\0.strtab\0.shstrtab\0.inte"..., 277, 138508034) = 277
    pread64(6, "`\3W\0\0\0\0\0\220\0W\0\0\0\0\0 \3W\0\0\0\0\0\360\323\31\5\0\0\0\0"..., 29612, 138474344) = 29612
    close(6)

Reply:
I also retested this on a somewhat large image (~550MB compressed), and the results were:

Syft main:
go run ./cmd/syft cloudbees/cloudbees-core-cm  36.64s user 14.22s system 93% cpu 54.441 total

This branch:
go run ./cmd/syft cloudbees/cloudbees-core-cm  38.29s user 15.00s system 100% cpu 53.254 total

... so an almost negligible difference from what I can tell at this point. 👍